### DISCUSSION TITLE
Is there a way to control the background color when using rmode=pad

### DISCUSSION THREAD

**Discussion by drgrieve** [SixLabors/ImageSharp.Web#376] [Q&A] [answered]
[Live image](https://portal.staging.auditcover.com/S3/Company/Io0hSU9f-kKTUWtphnc2Zw.png?h=168&w=1100&mode=pad&bgcolor=white)

h=168&w=1100&mode=pad&bgcolor=white

Seems to pad with the detected dominate color. I'd like to always pad with transparent or white.

![Example](https://github.com/user-attachments/assets/1fca658b-d45b-43b2-983f-3e11d2ba1fa2).

I understand there is an api to control bgcolor when doing a pad operation, but I'm unclear on how to hook into this side, I'm only using EnhanceParsedCommand

**Comment by JimBobSquarePants** [comment]
What is EnhanceParsedCommand? 

You want `rcolor` which takes the same values as `bgcolor` we don’t do any dominant color processing. If the image format supports alpha the default is transparent otherwise black

**Reply by drgrieve** [reply]
Thanks - rcolor has some impact and solves ones of my paths, but not another.

EnhanceParsedCommand was a miss paste was meant to be OnParseCommandsAsync

Also I don't get black for padding on this image, the padding is blue.

blue padding not back as expected => https://portal.staging.auditcover.com/S3/Company/Io0hSU9f-kKTUWtphnc2Zw.png?height=168&width=1100&rmode=pad

white padding when adding rcolor => https://portal.staging.auditcover.com/S3/Company/Io0hSU9f-kKTUWtphnc2Zw.png?height=168&width=1100&rmode=pad&rcolor=white

However rcolor is not working for transparent images, so adding rcolor breaks this path where for my usecase leaving rcolor off works

padding is #9FA0D7 not white => 
https://portal.staging.auditcover.com/S3/Company/KUATFtngR0iYRhA2CV900g.png?height=168&width=1100&rmode=pad&format=png&rcolor=white

padding is transparent => https://portal.staging.auditcover.com/S3/Company/KUATFtngR0iYRhA2CV900g.png?height=168&width=1100&rmode=pad&format=png

How do I specify a static command that works for both transparent and not transparent images with a padding color of either white or transparent?

(Note that I use shortcuts for commands, I'll edit them to be the native syntax)

**Reply by JimBobSquarePants** [reply, marked as answer]
Those images are palette-based PNGs which means they have a fixed color set. I would expect whoever provided these images have been pre-optimized for display on the web assuming any alignment would be done via CSS. 

ImageSharp by default will preserve any inbound color palette, in your case you are introducing transparency where there is none previously so the encoder will attempt to find a color in the palette which has the closes Euclidian distance to transparent. For example, blue is closer than white to transparent, so you get blue.

What you need to do in this case is to remove the inbound palette from the image metadata via [OnBeforeSaveAsync](https://docs.sixlabors.com/api/ImageSharp.Web/SixLabors.ImageSharp.Web.Middleware.ImageSharpMiddlewareOptions.html#SixLabors_ImageSharp_Web_Middleware_ImageSharpMiddlewareOptions_OnBeforeSaveAsync)

Something like the following...

```c#
ImageSharpMiddlewareOptions options = new()
{
    OnBeforeSaveAsync = static f =>
    {
        f.Image.Metadata.GetPngMetadata().ColorTable = null;
        return Task.CompletedTask;
    },
};
```
That will trigger the encoder to generate a new optimized palette on save which will include a transparent entry.

**Reply by drgrieve** [reply]
Thanks. That was very helpful in for me to move forward.

        private static Task OnSave(FormattedImage image)
        {
            if (image.Format is PngFormat)
            {
                image.Image.Metadata.GetPngMetadata().ColorTable = null;
            }
            return Task.CompletedTask;
        }
